api/agent/lb_agent.go

```diff
@@ -177,8 +185,41 @@ func (a *lbAgent) Submit(callI Call) error {
 	statsDequeue(ctx)
 	statsStartRun(ctx)
 
-	err = a.placer.PlaceCall(a.rp, ctx, call)
-	return a.handleCallEnd(ctx, call, err, true)
+	errPlace := make(chan error, 1)
+
+	call.w = &AckSyncResponseWriter{
+		origin: call.ResponseWriter(),
+		acked:  make(chan struct{}, 1),
+	}
+	isAckSync := call.Type == models.TypeAcksync
+	rw := call.w.(*AckSyncResponseWriter)
+
+	go a.spawnPlaceCall(ctx, call, errPlace, isAckSync)
+
+	for {
+		select {
+		case err := <-errPlace:
+			return err
+		case <-rw.acked:
+			// if it is an acksync we return immediately otherwise we ignore the ack
+			// If it is not an acksync we get the ack and we can use it to log info, stats etc.
+			if isAckSync {
+				return nil
+			}
+		}
+	}
+}
+
+func (a *lbAgent) spawnPlaceCall(ctx context.Context, call *call, errCh chan error, isAckSync bool) {
+	if isAckSync {
+		var cancel func()
+		ctx = common.BackgroundContext(ctx)
+		// We don't want this to run indefinetely we need to guard this context
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(60+call.Timeout)*time.Second)
```

Comment:
So, this 60 secs is related with the 30 secs we assume for async-ack placer config. Hardcoding this for now is OK, but let's add a comment about the math/rationale here. I think we said something like:

30 secs PlacerTimeout for AsyncAck + call.Timeout (inside container) + 30 sec headroom

to make sure we do not wait indefinitely, but also wait enough time for worst case.

Reply:
So, here the math is as follows:
a) placer could wait up to 30 secs, which means, it could wait 29.999 secs, then proceed to 'b'
b) container exec timeout
c) headroom: to account for delays in scheduling/network as well as docker-pull.

Now that I think about it, I think we do need to increase this to account for docker pull. I didn't think of docker pull duration before. We could keep it simple and set it to 360 secs + call.Timeout to make sure get close to 100% of cases as much as possible.